core/src/main/scala/chisel3/Data.scala

```diff
@@ -77,31 +85,60 @@ object SpecifiedDirection {
   * a node is bound (since higher-level specifications like Input and Output
   * can override directions).
   */
-sealed abstract class ActualDirection
+sealed abstract class ActualDirection(private[chisel3] val value: Byte)
 
 object ActualDirection {
 
+  // 0 is reserved for unset, no case object added because that would be an unnecessary API breakage.
+  private[chisel3] val Unset: Byte = 0
+
   /** The object does not exist / is empty and hence has no direction
     */
-  case object Empty extends ActualDirection
+  case object Empty extends ActualDirection(1)
 
   /** Undirectioned, struct-like
     */
-  case object Unspecified extends ActualDirection
+  case object Unspecified extends ActualDirection(2)
 
   /** Output element, or container with all outputs (even if forced)
     */
-  case object Output extends ActualDirection
+  case object Output extends ActualDirection(3)
 
   /** Input element, or container with all inputs (even if forced)
     */
-  case object Input extends ActualDirection
-
-  sealed abstract class BidirectionalDirection
-  case object Default extends BidirectionalDirection
-  case object Flipped extends BidirectionalDirection
+  case object Input extends ActualDirection(4)
+
+  sealed abstract class BidirectionalDirection(private[chisel3] val value: Byte)
+  case object Default extends BidirectionalDirection(5)
+  case object Flipped extends BidirectionalDirection(6)
```

Comment:
Can you add a comment here on why the index is starting at 5?

Should there be a check in the class that a lower value isn't in this field? Or is it safe enough b/c this is all private?

Reply:
Sure I'll add a comment, the numbering is a continuation of the numbers from ActualDirection above. These BidirectionalDirections are boxed by Bidirectional below, Bidirectional is an ActualDirection so this is basically flattening the representation while maintaining the public API. (I will say this in a comment 🙂)

Reply:
Done!